docs/source/contribution/threading.rst

```diff
@@ -0,0 +1,166 @@
+
+``daal::tls`` provides a local storage where each thread can accumulate its local results.
+Following code allocates memory that would store partial dot products for each thread:
+
+.. include:: ../includes/threading/dot-parallel-init-tls.rst
+
+``SafeStatus`` in this code denotes a thread-safe counterpart of the ``Status`` class.
+``SafeStatus`` allows to collect errors from all threads and report them to user using
+``detach()`` method as it will be shown later in the code.
+
+Checking the status right after the initialization code won't show the allocation errors,
+because oneTBB uses lazy evaluation and the lambda function passed to the constructor of the TLS
+is evaluated in the moment of the TLS's first use.
+
+Again, there are several options available in the threading layer of oneDAL to compute the partial
+dot product results at each thread.
+One of the options is to use the already mentioned ``daal::threader_for`` and blocking approach
+as shown here:
+
+.. include:: ../includes/threading/dot-parallel-partial-compute.rst
+
+To compute the final result it is requred to reduce TLS's partial results over all threads
+as shown here:
+
+.. include:: ../includes/threading/dot-parallel-reduction.rst
+
+Local memory of the threads should also be released when it is no longer needed.
+
+The complete parallel verision of dot product computations would look like:
+
+.. include:: ../includes/threading/dot-parallel.rst
+
+Static Work Scheduling
+**********************
+
+By default, oneTBB uses
+`dynamic work scheduling <https://oneapi-src.github.io/oneTBB/main/tbb_userguide/How_Task_Scheduler_Works.html>`_
+and work stealing.
+It means that two different runs of the same parallel loop can produce different
+mappings of the loop's iteration space to the available threads.
+This strategy is beneficial when it is difficult to estimate the amount of work performed
+by each iteration.
+
+In the cases when it is known that the iterations perform an equal amount of work, it
+might be beneficial to use predefined mapping of the loop's iterations to threads.
+This is what static work scheduling does.
+
+``daal::static_threader_for`` and ``daal::static_tls`` allow implementation of static
+work scheduling within oneDAL.
+
+Here is a variant of parallel dot product computation with static scheduling:
+
+.. include:: ../includes/threading/dot-static-parallel.rst
+
+Nested Parallelism
+******************
+
+oneDAL supports nested parallel loops.
+It is important to know that:
+
+    "when a parallel construct calls another parallel construct, a thread can obtain a task
+     from the outer-level construct while waiting for completion of the inner-level one."
+
+    -- `oneTBB documentation <https://www.intel.com/content/www/us/en/docs/onetbb/developer-guide-api-reference/2021-13/work-isolation.html>`_
+
+In practice, this means that a thread-local variable might unexpectedly
```

Comment:
Can you define what you mean by thread-local? I usually think of thread local data to mean that only a single thread can read or write the value to thread-local storage. The thread-local storage class `tls` is intended to be used for thread local storage, but it cannot do so in the case of nested parallelism. The only thread local storage which is guaranteed to be thread local is in the leaf nodes of a graph of threads.

So something along the lines of:
```
If we define thread-local data to be data that can only be read and written by a single thread, then any thread which spawns other threads in general cannot have thread-local variables. This is true for variables which are defined as thread-local. This means that great care is required when using thread-local storage constructs in nested parallel regions to avoid issues caused by concurrent data access and ownership (e.g. data corruption due to races; segmentation faults; and deadlocks)
```

Reply:
By thread-local data I mean the data that is managed by a particular thread.
The issue here is that there is a difference between the threads and logical tasks (loop's iterations here).
One thread can be assigned to execute several iterations of the loop.

In case of nested loops the specifics of oneTBB is that when the thread executes an outer iteration `i` of a nested loop it becomes a so called waiting thread while it waits for a completion of the inner parallel loop. And then this waiting thread can be assigned to perform another iteration `j` of an outer loop.
So, it might first complete the iteration `j` of the outer loop which will change the state of thread-local values, and then return back to iteration `i` .

I tried to reflect this behavior in the documentation's text.

In general I do not agree that the thread that spawns other threads cannot have thread-local data. But I agree that thread-local data should be used with great care in those cases.